### PR TITLE
add APIService conditions

### PR DESCRIPTION
**Pull request by deads2k**
Adds conditions to the APIServiceStatus struct and fixes up generators that appear to have slipped.

The first condition is "ServiceAvailable" which will provide the status currently derived in the discovery handler that decides about whether to expose the version in discovery.

@kubernetes/sig-api-machinery-pr-reviews @liggitt @ncdc

### CODE REVIEW

**Comment by deads2k** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by luxas** [comment]
@deads2k will you rebase this one anytime soon?
I'd love to have this working at HEAD ;)

File: staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/types.go [hunk 1]
```diff
@@ -61,8 +61,47 @@ type APIServiceSpec struct {
...
+	ConditionTrue    ConditionStatus = "True"
+	ConditionFalse   ConditionStatus = "False"
+	ConditionUnknown ConditionStatus = "Unknown"
+)
+
+// APIConditionConditionType is a valid value for APIServiceCondition.Type
+type APIServiceConditionType string
+
+const (
+	// ServiceAvailable indicates that the service exists and has endpoints
```
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/types.go]: Given other changes: "indicates that the service exists and is reachable"

**Comment by deads2k** [comment]
@liggitt @lavalamp second commit has alpha API changes for status.

File: staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/types.go [hunk 2]
```diff
@@ -61,8 +61,47 @@ type APIServiceSpec struct {
...
+type ConditionStatus string
+
+// These are valid condition statuses. "ConditionTrue" means a resource is in the condition;
+// "ConditionFalse" means a resource is not in the condition; "ConditionUnknown" means kubernetes
+// can't decide if a resource is in the condition or not. In the future, we could add other
+// intermediate conditions, e.g. ConditionDegraded.
+const (
+	ConditionTrue    ConditionStatus = "True"
+	ConditionFalse   ConditionStatus = "False"
+	ConditionUnknown ConditionStatus = "Unknown"
```
Comment (liggitt) [inline staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/types.go]: Are missing conditions treated as `Unknown`, or is that decided per condition type?

**Comment by liggitt** [comment]
assuming this is mostly replicating conditions from other API objects

**Comment by deads2k** [comment]
> assuming this is mostly replicating conditions from other API objects

correct.

File: staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/types.go (hunk 2 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/types.go]: > Are missing conditions treated as Unknown, or is that decided per condition type?

There is no guarantee that every condition will be present and there is no defaulting on these conditions..  How a client uses the data is up to them.  There is a difference between "Unknown" and `<missing>`, so hopefully clients know which one they want to test for.

**Comment by deads2k** [comment]
Comments addressed.

**Comment by deads2k** [comment]
@k8s-bot cvm gce e2e test this

**Comment by luxas** [comment]
ping @liggitt @ncdc 
It would be great to get this in ASAP :)

File: staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/types.go [hunk 3]
```diff
@@ -63,8 +63,45 @@ type APIServiceSpec struct {
...
+)
+
+// APIConditionConditionType is a valid value for APIServiceCondition.Type
+type APIServiceConditionType string
+
+const (
+	// Available indicates that the service exists and is reachable
```
Comment (liggitt) [inline staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/types.go]: does this single condition type encompass indicating the resource/group/version was not claimed by another API service, and that the backend service was reachable?

**Comment by liggitt** [comment]
one question, no comments otherwise

File: staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/types.go (hunk 3 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/types.go]: > does this single condition type encompass indicating the resource/group/version was not claimed by another API service, and that the backend service was reachable?

Request from @smarterclayton.  I think it lacks granularity, but I need something more than I need my preferred thing and as a summary it's an ok name.

**Comment by deads2k** [comment]
top level stuff is straight generated. approving.

**Comment by luxas** [comment]
LGTM, thank you @deads2k!

**Pull request merged**